File: 41. First Missing Positive.py
class Solution:
    def firstMissingPositive(self, nums: List[int]) -> int:
        i = 0
        
        while i<len(nums):
            cur = nums[i]
            
            while (cur>0 and cur<=len(nums)) and cur!=nums[cur-1] and cur-1!=i:
                nums[cur-1], cur = cur, nums[cur-1]

            nums[i] = cur
            i += 1

                
        for i, n in enumerate(nums):
            if n!=i+1:
                return i+1
        
        return len(nums)+1

File: 41. First Missing Positive.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var firstMissingPositive = function(nums) {
    let i=0;

    while (i<nums.length){
        let cur = nums[i];

        while ((cur>0 && cur<=nums.length) 
              && cur!==nums[cur-1] && cur-1!==i){
            [nums[cur-1], cur] = [cur, nums[cur-1]];
        }
        nums[i] = cur;
        i += 1;
    }
    for (let i=0;i<nums.length;i++){
        if (nums[i]!=i+1){
            return i+1;
        }
    }
    return nums.length+1;
};